portality/lib/formulaic.py

```diff
@@ -266,6 +268,20 @@ def default_field_template(self):
     def default_group_template(self):
         return self._definition.get("templates", {}).get("default_group")
 
+    def list_fields_in_order(self):
+        fieldlist = []
+        for fs in self.fieldsets():
+            for field in fs.fields():
+                if field.group_subfields():
+                    for sf in field.group_subfields():
+                        fieldlist.append(sf)
+                else:
+                    fieldlist.append(field)
+        return fieldlist
+
+    def lang_to_country(self, lang):
+        return datasets.language_for(lang).name
```

Comment:
What happens if the `lang` is not found? Presumably `.name` causes an exception.

Also, are we converting the language to a country, or are we converting the language to the language name?

In any case, this feels to form-content-specific to be in formulaic, which ought to be generic.  As this is just a wrapper for a call to the datasets module, the calling code should just use the datasets module directly, imo.

Reply:
it actually should convert language to the language name for language field and country code for the country for the country field. I am just wondering whether the template really should cause datasets module directly. It is not really form-specific. It is the same conversion for every form (although used for diff table). Not creating the wrapper for datasets module in formulaic will require passing that module as a param to each form template - not ideal either... What do you think? @richard-jones

Reply:
I'm mid-way on this - my first instict was that a thin wrapper around datasets wasn't necessary, but I agree that exposing that entire lib to the template is also clunky.

Would it be better to define this as a template filter? (not sure if you can do it on blueprints instead of app) - see https://github.com/DOAJ/doaj/blob/develop/portality/app.py#L182 for examples.

Reply:
@Steven-Eardley after call with Richard the datasets module is now exposed to jinja templates by setup in core.py. What do you think about it?

Reply:
@Steven-Eardley's approach with a template filter is the other way to do this, but I think since we need a few functions from `datasets` making it available as a global to the templating system makes the most sense now.  There are probably a few other places in the templates where we could use this, too.